File: app/src/main/java/pl/vemu/zsme/newsFragment/NewsFragment.java
package pl.vemu.zsme.newsFragment;

import android.content.Context;
import android.net.ConnectivityManager;
import android.net.Network;
import android.net.NetworkRequest;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.SearchView;
import androidx.databinding.DataBindingUtil;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.RecyclerView;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;

import java.util.ArrayList;
import java.util.Objects;

import pl.vemu.zsme.BaseAdapter;
import pl.vemu.zsme.R;
import pl.vemu.zsme.databinding.FragmentNewsBinding;

public class NewsFragment extends Fragment implements SwipeRefreshLayout.OnRefreshListener {

    private BaseAdapter adapter;
    private FragmentNewsBinding binding;
    private RecyclerView.OnScrollListener scrollListener;
    private SearchView searchView;
    private String author;
    private NewsFragmentVM viewmodel;

    public NewsFragment() {
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        binding = DataBindingUtil.inflate(inflater, R.layout.fragment_news, container, false);
        viewmodel = new ViewModelProvider(this).get(NewsFragmentVM.class);
        binding.setLifecycleOwner(this);
        binding.setViewmodel(viewmodel);
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        author = NewsFragmentArgs.fromBundle(requireArguments()).getAuthor();

        setupRecyclerView();

        setHasOptionsMenu(true);

        binding.refresh.setOnRefreshListener(this);
        binding.refresh.setColorSchemeResources(R.color.colorPrimary);

        setupNetwork();
    }

    private void setupNetwork() {
        ConnectivityManager connectivityManager = (ConnectivityManager) requireContext().getSystemService(Context.CONNECTIVITY_SERVICE);
        NetworkRequest networkRequest = new NetworkRequest.Builder().build();
        if (connectivityManager.getActiveNetwork() == null)
            Toast.makeText(getContext(), "Brak połaczenia z internetem", Toast.LENGTH_LONG).show();
        connectivityManager.registerNetworkCallback(networkRequest, networkCallback);
    }

    private void setupRecyclerView() {
        adapter = new BaseAdapter(R.layout.item_news, new ArrayList<>(Objects.requireNonNull(viewmodel.getList().getValue())));
        adapter.setHasStableIds(true);
        binding.recyclerView.setAdapter(adapter);
        scrollListener = new NewsScrollListener(viewmodel, new Queries.Page());
        binding.recyclerView.addOnScrollListener(scrollListener);
        viewmodel.getList().observe(getViewLifecycleOwner(), newsItems -> {
            adapter.setList(new ArrayList<>(newsItems));
            adapter.notifyDataSetChanged();
        });
    }

    private final ConnectivityManager.NetworkCallback networkCallback = new ConnectivityManager.NetworkCallback() {
        @Override
        public void onAvailable(@NonNull Network network) {
            if (author == null && viewmodel.getList().getValue() != null && viewmodel.getList().getValue().isEmpty())
                downloadFirstNews();
        }
    };

    private void downloadFirstNews() {
        viewmodel.clearList();
        viewmodel.downloadNews(new Queries.Page());
    }

    @Override
    public void onCreateOptionsMenu(@NonNull Menu menu, @NonNull MenuInflater inflater) {
        inflater.inflate(R.menu.menu_search, menu);
        searchView = (SearchView) menu.findItem(R.id.app_bar_search).getActionView();
        searchView.setOnQueryTextListener(new NewsQueryTextListener(viewmodel, binding.recyclerView));
        searchView.setOnCloseListener(() -> {
            searchView.onActionViewCollapsed();
            binding.recyclerView.clearOnScrollListeners();
            binding.recyclerView.addOnScrollListener(scrollListener);
            downloadFirstNews();
            return true;
        });
        if (author != null) {
            searchView.onActionViewExpanded();
            searchView.setQuery(author, true);
        }
    }

    @Override
    public void onRefresh() {
        if (!"".contentEquals(searchView.getQuery())) {
            searchView.setQuery(searchView.getQuery(), true);
        } else downloadFirstNews();
    }
}

File: app/src/main/java/pl/vemu/zsme/newsFragment/NewsFragment.kt
package pl.vemu.zsme.newsFragment

import android.content.Context
import android.net.ConnectivityManager
import android.net.Network
import android.net.NetworkRequest
import android.os.Bundle
import android.util.Log
import android.view.*
import android.widget.Toast
import androidx.appcompat.widget.SearchView
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.lifecycle.lifecycleScope
import androidx.navigation.fragment.navArgs
import androidx.recyclerview.widget.RecyclerView
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout.OnRefreshListener
import com.google.android.material.snackbar.Snackbar
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.flow.collect
import pl.vemu.zsme.R
import pl.vemu.zsme.SimpleAdapter
import pl.vemu.zsme.State
import pl.vemu.zsme.databinding.FragmentNewsBinding
import pl.vemu.zsme.model.PostModel

@AndroidEntryPoint
class NewsFragment : Fragment(), OnRefreshListener {

    private var _binding: FragmentNewsBinding? = null
    private val binding get() = _binding!!
    private var scrollListener: RecyclerView.OnScrollListener? = null //TODO remove
    private var searchView: SearchView? = null //TODO remove
    private val args: NewsFragmentArgs by navArgs()
    private val viewmodel: NewsFragmentVM by viewModels()

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        _binding = FragmentNewsBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onDestroy() {
        super.onDestroy()
        _binding = null
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        setupRecyclerView()
        //        setHasOptionsMenu(true) TODO remvoe
        binding.refresh.apply {
            setOnRefreshListener(this@NewsFragment)
            setColorSchemeColors(resources.getColor(R.color.colorPrimary, null))
            setProgressBackgroundColorSchemeColor(resources.getColor(R.color.swipeBackground, null))
        }
        setupNetwork()
    }

    private fun setupNetwork() {
        val cm = context?.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
        val networkRequest = NetworkRequest.Builder().build()
        if (cm.activeNetwork == null) Toast.makeText(context, "Brak połaczenia z internetem", Toast.LENGTH_LONG).show()
        cm.registerNetworkCallback(networkRequest, object : ConnectivityManager.NetworkCallback() {
            override fun onAvailable(network: Network) {
                //                viewmodel.fetchPosts()
            }
        })
    }

    private fun setupRecyclerView() {
        val adapter = SimpleAdapter<PostModel>(R.layout.item_news, emptyList()).apply {
            setHasStableIds(true)
        }
        binding.recyclerView.adapter = adapter
        //        scrollListener = NewsScrollListener(viewmodel, Queries.Page())
        //        binding.recyclerView.addOnScrollListener(scrollListener)
        lifecycleScope.launchWhenStarted {
            viewmodel.posts.collect {
                when (it) {
                    is State.Success -> {
                        binding.refresh.isRefreshing = false
                        adapter.list = it.data
                        adapter.notifyDataSetChanged()
                    }
                    is State.Loading -> binding.refresh.isRefreshing = true
                    is State.Error -> {
                        binding.refresh.isRefreshing = false
                        it.throwable.message?.let { message -> Log.e("ZSMEDebug", message) }
                        Snackbar.make(binding.root, R.string.error, Snackbar.LENGTH_SHORT).show()
                    }
                }
            }
        }
    }

    /*private fun downloadFirstNews() {
        viewmodel.clearList()
        viewmodel.downloadNews(Queries.Page())
    }*/

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        inflater.inflate(R.menu.menu_search, menu)
        searchView = menu.findItem(R.id.app_bar_search).actionView as SearchView
        searchView!!.setOnQueryTextListener(NewsQueryTextListener(viewmodel, binding.recyclerView))
        searchView!!.setOnCloseListener {
            searchView!!.onActionViewCollapsed()
            binding.recyclerView.clearOnScrollListeners()
            binding.recyclerView.addOnScrollListener(scrollListener!!)
            //            downloadFirstNews()
            true
        }
        if (args.author != null) {
            searchView!!.onActionViewExpanded()
            searchView!!.setQuery(args.author, true)
        }
    }

    override fun onRefresh() {
        /*if (!"".contentEquals(searchView!!.query)) {
            searchView!!.setQuery(searchView!!.query, true)
        } else downloadFirstNews()*/
        viewmodel.fetchPosts()
    }
}